[#5647] Advance bastion orders based on calendar

arbron opened the pull request:
Automatically advance bastion orders based on when new days occur using the built-in calendar. Allows orders to complete without needing a bastion turn to be triggered. Each facility with an order in progress stores the time it was last updated and that is used to advance or complete the order when time passes.

 ### Todo
- [ ] Prompt for bastion turns on regular interval

arbron committed:
[#5647] Advance bastion orders based on calendar

Automatically advance bastion orders based on when new days occur
using the built-in calendar. Allows orders to complete without
needing a bastion turn to be triggered. Each facility with an
order in progress stores the time it was last updated and that is
used to advance or complete the order when time passes.

 ### Todo
- [ ] Prompt for bastion turns on regular interval

Fyorl commented on module/data/settings/bastion-setting.mjs:
```diff
@@ -24,4 +24,31 @@ export default class BastionSetting extends foundry.abstract.DataModel {
+
+  /**
+   * Cached version of the minimum level required to have a bastion.
+   * @param {number}
```
```suggestion
   * @type {number}
```

Fyorl commented on module/documents/actor/bastion.mjs:
```diff
@@ -29,64 +30,84 @@ export default class Bastion {
-      await facility.update({ "system.progress.order": "" });
-      if ( type.value === "basic" ) return {}; // Basic facilities do nothing.
-      return { order: "maintain" }; // Special facilities are considered to have been issued the maintain order.
+      const updates = { "system.progress.order": "" };
```
Since `handleTimePassage` is sensitive to whether the delta is empty or not, we can return an empty delta in some cases. An idle facility already has `order: ""` for example.


```suggestion
      const updates = progress.order ? { "system.progress.order": "" } : {};
```

Fyorl commented on module/documents/actor/bastion.mjs:
```diff
@@ -29,64 +30,84 @@ export default class Bastion {
+      if ( days > 0 ) duration = days;
     }
 
     const newProgress = Math.min(progress.value + duration, progress.max);
```
It's possible for `duration` to still be `null` here.


```suggestion
    duration ??= 0;
    const newProgress = Math.min(progress.value + duration, progress.max);
```

Fyorl commented on module/data/settings/bastion-setting.mjs:
```diff
@@ -24,4 +24,31 @@ export default class BastionSetting extends foundry.abstract.DataModel {
+   * Cached version of the minimum level required to have a bastion.
+   * @param {number}
+   */
+  static #threshold;
```
What's the earliest we might cache this? Is it possible a module updates `CONFIG.DND5E.facilities.advancement` later?

arbron replied:
It is cached either when the first character sheet is opened or when time is advanced past midnight, so both would generally be past the `ready` hook so it shouldn't be an issue.

Fyorl commented on module/documents/actor/bastion.mjs:
```diff
@@ -29,64 +30,84 @@ export default class Bastion {
+
+    // Calculate duration if automatic
+    if ( !duration && progress.updated ) {
+      const days = CalendarData5e.dayDifference(
```
Is it not sufficient to pass `timePassageData.midnights` as the `duration`? Do we need to store an `updated` timestamp on the individual facilities?

arbron replied:
My intention here was to avoid double-triggering if time was reversed and then advanced again. Since we don't do that for recovery I'm fine not doing the extra work here.

Fyorl replied:
Right, we would need a timestamp to handle that, makes sense. Either that or we make recovery/progress reverse when time reverses I suppose.

Fyorl committed:
Do not wipe progress when a facility is disabled. Allow progress to continue once a facility is repaired.

Fyorl reviewed (changes requested):
Thanks for doing this. I think we are able to model Bastions much better now with these changes.

As part of my review, I went over the rules again to re-familiarise myself. That's why you'll see I made a small update to change the behaviour of disabled facilities. Briefly, the rules say that a disabled facility cannot be *used*, but they don't say progress is interrupted. That was a logical leap that isn't actually mentioned anywhere in the rules that I could tell. So without a rule that says 'a disabled facility loses its progress', there's no reason to assume that's the case.

But it revealed an issue in the PR. Because `advanceTurn` is called every day now, we would actually repair a facility in one day rather than one bastion turn.

So for the most part, this PR is actually much more correct than what we had before. All durations in the bastion rules are specified in days, they are totally disconnected from bastion turns. A bastion turn is something a DM calls for, at their discretion, which the book *recommends* is every 7 days.

A bastion turn elapsing does very few things: (a) It issues the Maintain order to the whole Bastion for each character that is not present at the Bastion or unable to communicate with it. This prompts a roll on the events table. (b) It causes damaged facilities to be repaired.

So what we need, I think, is a clear distinction between *day-quantised* events, and *turn-quantised* events. We have the day-quantised events down pretty much already, it's just the turn-quantised events that need some additional work in this PR then.

I would propose then, if you have the calendar enabled, that the 'Advance Bastion Turn' button becomes a 'Maintain Bastion' button instead. It issues the Maintain order to all idle facilities and (later) it can roll on the events table, possibly triggering an attack, which we already have code for. It will also cause damaged facilities to be repaired.

If you don't have the calendar enabled, then the button just works exactly as it did before and it rolls progress + bastion turn all together. That means our `// TODO: Should this advance game.time` is 'no', and we can delete that comment.

arbron committed:
[#5647] Only repair on full turn, change button, other fixes

- Repairing facilities only completes when GM uses the "Bastion
  Turn" button, not through time advancing
- Bastion turn button now behaves differently depending on whether
  the calendar is enabled, only triggering turn events rather than
  advancing progress if disabled
- Last updated time is no longer stored for each facility

Fyorl commented on module/data/settings/bastion-setting.mjs:
```diff
@@ -24,4 +24,31 @@ export default class BastionSetting extends foundry.abstract.DataModel {
+      const { basic, special } = CONFIG.DND5E.facilities.advancement;
+      BastionSetting.#threshold = Math.min(...Object.keys(basic), ...Object.keys(special));
+    }
+    return this.enabled && (actor.system.details?.level >= BastionSetting.#threshold);
```
```suggestion
    return this.enabled && system.isCharacter && (actor.system.details?.level >= BastionSetting.#threshold);
```

arbron committed:
[#5647] Only check bastions on character actors
